File: test/unit/recurrence_test.dart
import 'package:decimal/decimal.dart';
import 'package:fintech_os/domain/entities/recurring_rule.dart';
import 'package:fintech_os/domain/entities/transaction.dart';
import 'package:fintech_os/domain/services/recurrence_calculator.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const calc = RecurrenceCalculator();

  RecurringRule rule(Frequency freq, DateTime nextRun) => RecurringRule(
        id: 'r',
        vaultId: 'v',
        amount: Decimal.parse('1000'),
        type: TxnType.expense,
        categoryId: 'c',
        frequency: freq,
        nextRun: nextRun,
      );

  group('next()', () {
    test('daily/weekly/monthly/yearly', () {
      final base = DateTime(2026, 1, 15);
      expect(calc.next(base, Frequency.daily), DateTime(2026, 1, 16));
      expect(calc.next(base, Frequency.weekly), DateTime(2026, 1, 22));
      expect(calc.next(base, Frequency.monthly), DateTime(2026, 2, 15));
      expect(calc.next(base, Frequency.yearly), DateTime(2027, 1, 15));
    });
  });

  group('materialize() catch-up', () {
    test('no occurrences when nextRun is in the future', () {
      final r = rule(Frequency.monthly, DateTime(2026, 12, 1));
      final res = calc.materialize(r, DateTime(2026, 6, 1));
      expect(res.due, isEmpty);
      expect(res.newNextRun, DateTime(2026, 12, 1));
    });

    test('catches up missed monthly runs', () {
      // nextRun Jan 1, now Apr 15 → Jan, Feb, Mar, Apr = 4 occurrences.
      final r = rule(Frequency.monthly, DateTime(2026, 1, 1));
      final res = calc.materialize(r, DateTime(2026, 4, 15));
      expect(res.due.length, 4);
      expect(res.newNextRun, DateTime(2026, 5, 1));
    });

    test('single due occurrence advances by one period', () {
      final r = rule(Frequency.weekly, DateTime(2026, 6, 1));
      final res = calc.materialize(r, DateTime(2026, 6, 3));
      expect(res.due.length, 1);
      expect(res.newNextRun, DateTime(2026, 6, 8));
    });
  });
}

File: webapp/src/domain/recurrence.test.ts
import { describe, test, expect } from 'vitest';
import { advance, materialize } from './recurrence';
import type { RecurringRule } from '@/lib/types';

const d = (iso: string) => new Date(iso).getTime();
const day = (t: number) => new Date(t).getDate();

const rule = (over: Partial<RecurringRule>): RecurringRule => ({
  id: 'r1', vaultId: 'v1', amount: '70000', type: 'income',
  categoryId: 'c1', frequency: 'monthly', nextRun: d('2026-09-07T00:00:00'), ...over,
});

describe('advance', () => {
  // The scheduled day of month is the intent, and it has to survive a late
  // posting. A salary due on the 7th that is paid on the 10th is still a rule
  // about the 7th — stepping from the payment date would move every future
  // occurrence to the 10th permanently.
  test('a late payment does not drag the schedule when stepped from the schedule', () => {
    const scheduled = d('2026-09-07T00:00:00');
    const paidLate = d('2026-09-10T00:00:00');
    expect(day(advance(scheduled, 'monthly'))).toBe(7);
    expect(day(advance(paidLate, 'monthly'))).toBe(10);
  });

  test('monthly clamps into a short month and returns to the intended day', () => {
    const jan31 = d('2026-01-31T00:00:00');
    const feb = advance(jan31, 'monthly');
    expect(day(feb)).toBe(28);
    expect(day(advance(feb, 'monthly'))).toBe(28);
    // Stepping from the original 31st reaches March the 31st, not the 3rd of
    // April — the overflow bug this clamping exists to prevent.
    expect(day(advance(d('2026-03-31T00:00:00'), 'monthly'))).toBe(30);
  });

  test('daily, weekly and yearly step as expected', () => {
    expect(advance(d('2026-09-07T00:00:00'), 'daily')).toBe(d('2026-09-08T00:00:00'));
    expect(advance(d('2026-09-07T00:00:00'), 'weekly')).toBe(d('2026-09-14T00:00:00'));
    expect(advance(d('2026-09-07T00:00:00'), 'yearly')).toBe(d('2027-09-07T00:00:00'));
  });
});

describe('materialize', () => {
  // This is what makes the batch run all-or-nothing, and why posting one rule
  // needed its own path: a rule three days overdue reports one missed run, and
  // running the batch posts that one for EVERY overdue rule at once.
  test('reports the runs a rule has missed', () => {
    const { runs, nextRun } = materialize(rule({}), d('2026-09-10T00:00:00'));
    expect(runs).toEqual([d('2026-09-07T00:00:00')]);
    expect(day(nextRun)).toBe(7);
    expect(new Date(nextRun).getMonth()).toBe(9); // October
  });

  test('catches up several missed periods at once', () => {
    const { runs } = materialize(
      rule({ nextRun: d('2026-06-07T00:00:00') }),
      d('2026-09-10T00:00:00'),
    );
    expect(runs).toHaveLength(4); // Jun, Jul, Aug, Sep
  });

  test('a rule not yet due produces no runs and keeps its date', () => {
    const r = rule({});
    const { runs, nextRun } = materialize(r, d('2026-09-01T00:00:00'));
    expect(runs).toEqual([]);
    expect(nextRun).toBe(r.nextRun);
  });
});
